Extension/src/Debugger/configurationProvider.ts

```diff
@@ -593,13 +593,100 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
+    }
+
+    public async writeDebugConfig(config: vscode.DebugConfiguration, folder?: vscode.WorkspaceFolder): Promise<void> {
+        const settings: OtherSettings = new OtherSettings();
+        const launchJsonPath: string | undefined = this.getLaunchJsonPath();
+
+        if (this.checkDebugConfigExists(config.name, folder)) {
+            if (launchJsonPath) {
+                const doc: vscode.TextDocument = await vscode.workspace.openTextDocument(launchJsonPath);
+                if (doc) {
+                    vscode.window.showTextDocument(doc);
+                }
+            }
+            return;
+        }
+        const rawLaunchJson: any = await this.getRawLaunchJson();
+        if (!rawLaunchJson.configurations) {
+            rawLaunchJson.configurations = new Array();
+        }
+        if (!rawLaunchJson.version) {
+            rawLaunchJson.version = "2.0.0";
+        }
+
+        config.detail = undefined;
+        config.existing = undefined;
+        config.isDefault = undefined;
+        rawLaunchJson.configurations.push(config);
+
+        if (!launchJsonPath) {
+            throw new Error("Failed to get tasksJsonPath in checkBuildTaskExists()");
+        }
+
+        await util.writeFileText(launchJsonPath, jsonc.stringify(rawLaunchJson, null, settings.editorTabSize));
+        await vscode.workspace.openTextDocument(launchJsonPath);
+        const doc: vscode.TextDocument = await vscode.workspace.openTextDocument(launchJsonPath);
+        if (doc) {
+            vscode.window.showTextDocument(doc);
+        }
+    }
+
+    public async addDebugConfiguration(textEditor: vscode.TextEditor): Promise<void> {
+        const folder: vscode.WorkspaceFolder | undefined = vscode.workspace.getWorkspaceFolder(textEditor.document.uri);
+        if (!folder) {
+            return;
+        }
+        const selectedConfig: vscode.DebugConfiguration | undefined = await this.selectConfiguration(textEditor, true);
+        if (!selectedConfig) {
+            Telemetry.logDebuggerEvent(DebuggerEvent.launchPlayButton, { "debugType": "AddConfigurationOnly", "folderMode": folder ? "folder" : "singleFile", "cancelled": "true" });
+            return; // User canceled it.
+        }
+
+        if (this.isClConfiguration(selectedConfig.name) && this.showErrorIfClNotAvailable(selectedConfig.name)) {
+            return;
+        }
+        // Write preLaunchTask into tasks.json file.
+        if (selectedConfig.preLaunchTask) {
+            await cppBuildTaskProvider.writeBuildTask(selectedConfig.preLaunchTask);
+        }
+        // Write debug configuraion in launch.json file.
+        await this.writeDebugConfig(selectedConfig, folder);
+
+    }
+
     public async buildAndRun(textEditor: vscode.TextEditor): Promise<void> {
         // Turn off the debug mode.
         return this.buildAndDebug(textEditor, false);
     }
 
     public async buildAndDebug(textEditor: vscode.TextEditor, debugModeOn: boolean = true): Promise<void> {
```

Comment:
`addDebugConfiguration` and `buildAndDebug` has some duplicate code. Place duplicate code in helper function to reuse in both`addDebugConfiguration` and `buildAndDebug`?

Reply:
the telemetry part was not duplicate so I kept it there, but the checkng for ClConfiguration was duplicate and I moved it to the higher function.